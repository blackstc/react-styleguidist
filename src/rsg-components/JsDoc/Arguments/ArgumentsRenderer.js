import React from 'react';
import PropTypes from 'prop-types';
import Styled from 'rsg-components/Styled';
import Markdown from 'rsg-components/Markdown';
import Code from 'rsg-components/Code';
import Group from 'react-group';

export const styles = ({ type, name }) => ({
	type: {
		fontSize: 'inherit',
		color: type,
	},
	name: {
		fontSize: 'inherit',
		color: name,
	},
	argumentsBlock: {
		marginTop: 6,
		fontSize: 'inherit',
	},
	argumentsHeader: {
		fontWeight: 'bold',
		fontSize: 'inherit',
	},
});

export function ArgumentsRenderer({ classes, tags }) {
	if (!tags || (!tags.arg && !tags.argument && !tags.param)) {
		return null;
	}

	function renderParams(params) {
		return (
			params.map((param) => {
				return (
					<Group key={`${param.name}`}>
						<Code className={classes.name}>{param.name}</Code>
						{param.type && ':'}
						{param.type && <Code className={classes.type}>{param.type.name}</Code>}
						{param.description && <Markdown text={`â€” ${param.description}`} inline />}
					</Group>
				);
			})
		);
	}

	return (
		<div className={classes.argumentsBlock}>
			<div className={classes.argumentsHeader}>Arguments</div>
			{ tags.param && renderParams(tags.param) }
			{ tags.arg && renderParams(tags.arg) }
			{ tags.argument && renderParams(tags.argument) }
		</div>
	);
}

ArgumentsRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	tags: PropTypes.shape({
		arg: PropTypes.array,
		argument: PropTypes.array,
		param: PropTypes.array,
	}).isRequired,
};

export default Styled(styles)(ArgumentsRenderer);
