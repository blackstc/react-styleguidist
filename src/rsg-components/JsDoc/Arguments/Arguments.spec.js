import React from 'react';
import { ArgumentsRenderer } from './ArgumentsRenderer';

describe('Arguments', () => {
	function createTestParameter(title, name = 'test', description = 'Test description', type = { name: 'string' }) {
		return {
			title,
			name,
			description,
			type,
		};
	}

	const mockParam = {
		param: [createTestParameter('param', 'one'), createTestParameter('param', 'two')],
	};

	const mockArg = {
		arg: [createTestParameter('arg')],
	};

	const mockArgument = {
		argument: [createTestParameter('argument')],
	};

	it('renderer should render Arguments for one @param tag', () => {
		const tags = {
			param: [mockParam.param[0]],
		};

		const actual = render(
			<ArgumentsRenderer classes={{}} tags={tags} />
		);

		expect(actual).toMatchSnapshot();
	});

	it('renderer should render Arguments for two @param tags', () => {
		const actual = render(
			<ArgumentsRenderer classes={{}} tags={mockParam} />
		);

		expect(actual).toMatchSnapshot();
	});

	it('renderer should render Arguments for @arg tag', () => {
		const actual = render(
			<ArgumentsRenderer classes={{}} tags={mockArg} />
		);

		expect(actual).toMatchSnapshot();
	});

	it('renderer should render Arguments for @argument tag', () => {
		const actual = render(
			<ArgumentsRenderer classes={{}} tags={mockArgument} />
		);

		expect(actual).toMatchSnapshot();
	});

	it('renderer should render Arguments for mixed tags', () => {
		const mixed = {
			...mockParam,
			...mockArg,
			...mockArgument,
		};

		const actual = render(
			<ArgumentsRenderer classes={{}} tags={mixed} />
		);

		expect(actual).toMatchSnapshot();
	});

	it('renderer should still render for incomplete documentation', () => {
		const tags = {
			param: [
				createTestParameter('param', 'testNoDescription', null),
				createTestParameter('param', 'testNoType', 'Test description', null),
			],
		};

		const actual = render(
			<ArgumentsRenderer tags={tags} classes={{}} />
		);

		expect(actual).toMatchSnapshot();
	});

	it('renderer should render empty for empty documentation', () => {
		const actual = shallow(
			<ArgumentsRenderer tags={{}} classes={{}} />
		);

		expect(actual.node).toBe(null);
	});
});

