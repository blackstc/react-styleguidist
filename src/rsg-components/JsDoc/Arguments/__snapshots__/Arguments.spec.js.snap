// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arguments renderer should render Arguments for @arg tag 1`] = `
<div>
  <div>
    Arguments
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        test
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
</div>
`;

exports[`Arguments renderer should render Arguments for @argument tag 1`] = `
<div>
  <div>
    Arguments
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        test
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
</div>
`;

exports[`Arguments renderer should render Arguments for mixed tags 1`] = `
<div>
  <div>
    Arguments
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        one
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        two
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        test
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        test
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
</div>
`;

exports[`Arguments renderer should render Arguments for one @param tag 1`] = `
<div>
  <div>
    Arguments
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        one
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
</div>
`;

exports[`Arguments renderer should render Arguments for two @param tags 1`] = `
<div>
  <div>
    Arguments
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        one
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        two
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
</div>
`;

exports[`Arguments renderer should still render for incomplete documentation 1`] = `
<div>
  <div>
    Arguments
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        testNoDescription
      </code>
    </span>
     : 
    <span>
      <code
        class="code-0-0"
      >
        string
      </code>
    </span>
  </div>
  <div>
    <span>
      <code
        class="code-0-0"
      >
        testNoType
      </code>
    </span>
     
    <span
      class="base-0-1"
    >
      — Test description
    </span>
  </div>
</div>
`;
